packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart

```diff
@@ -13,50 +13,138 @@ class AdaptedHeap {
   late final HeapStatistics stats = _heapStatistics(data);
 
   static HeapStatistics _heapStatistics(AdaptedHeapData data) {
-    final result = <String, HeapStatsRecord>{};
+    final result = <String, HeapClassStatistics>{};
     if (!data.isSpanningTreeBuilt) buildSpanningTree(data);
 
-    for (var object in data.objects) {
+    for (var i in Iterable.generate(data.objects.length)) {
+      final object = data.objects[i];
       final heapClass = object.heapClass;
 
-      // We do not show objects that will be garbage collected soon.
+      // We do not show objects that will be garbage collected soon or are
+      // native.
       if (object.retainedSize == null || heapClass.isSentinel) continue;
 
       final fullName = heapClass.fullName;
-      if (!result.containsKey(fullName)) {
-        result[fullName] = HeapStatsRecord(heapClass);
-      }
-      final stats = result[fullName]!;
-      stats.retainedSize += object.retainedSize ?? 0;
-      stats.shallowSize += object.shallowSize;
-      stats.instanceCount++;
+
+      final stats =
+          result.putIfAbsent(fullName, () => HeapClassStatistics(heapClass));
+      stats.countInstance(data, i);
     }
-    return HeapStatistics(result);
+
+    return HeapStatistics(result)..seal();
   }
 }
 
 class HeapStatistics {
-  HeapStatistics(this.recordsByClass);
+  HeapStatistics(this.statsByClassName);
 
-  /// Maps full class name to stats record of this class.
-  final Map<String, HeapStatsRecord> recordsByClass;
-  late final List<HeapStatsRecord> records =
-      recordsByClass.values.toList(growable: false);
+  /// Maps full class name to statistics of this class.
+  final Map<String, HeapClassStatistics> statsByClassName;
+  late final List<HeapClassStatistics> classStats =
+      statsByClassName.values.toList(growable: false);
+
+  /// Mark the object as deeply immutable.
+  ///
+  /// There is no strong protection from mutation, just some asserts.
+  void seal() {
+    for (var stats in classStats) {
+      stats.seal();
+    }
+  }
 }
 
-class HeapStatsRecord {
-  HeapStatsRecord(this.heapClass);
+class HeapClassStatistics {
+  HeapClassStatistics(this.heapClass)
+      : _isSealed = false,
+        total = SizeOfClassSet(),
+        sizeByRetainingPath = <ClassOnlyHeapPath, SizeOfClassSet>{};
+
+  HeapClassStatistics.negative(HeapClassStatistics other)
+      : _isSealed = true,
+        heapClass = other.heapClass,
+        total = SizeOfClassSet.negative(other.total),
+        sizeByRetainingPath = other.sizeByRetainingPath
+            .map((key, value) => MapEntry(key, SizeOfClassSet.negative(value)));
+
+  HeapClassStatistics.subtract(
+    HeapClassStatistics left,
+    HeapClassStatistics right,
+  )   : assert(left.heapClass.fullName == right.heapClass.fullName),
+        _isSealed = true,
```

Comment:
Is sealing used to make sure that we never count instances that have been sealed?
If so could we add some JSDoc explaining what the sealing is needed for or how it is being used?